File: apps/web/app/nyushukkin.ts
import * as v from "valibot";

export const kinds = ["支出", "収入"] as const;
export type Kind = (typeof kinds)[number];

export type Nyushukkin = {
  id: string;
  kind: Kind;
  amount: number;
  date: string;
  memo: string;
};

export type NyushukkinInput = {
  kind: string;
  amount: string;
  date: string;
  memo: string;
};

export type NyushukkinResult<T> = { ok: true; value: T } | { ok: false; error: string };

const datePattern = /^\d{4}-\d{2}-\d{2}$/;
const amountPattern = /^\d+$/;
const kindError = "収入か支出を選んでください";
const amountError = "金額は1円以上の整数円です";
const dateError = "入出日は今日以前の日付です";

const kindSchema = v.picklist(kinds, kindError);
const amountInputSchema = v.pipe(
  v.string(),
  v.trim(),
  v.check((raw) => {
    if (!amountPattern.test(raw)) {
      return false;
    }
    const amount = Number(raw);
    return Number.isSafeInteger(amount) && amount >= 1;
  }, amountError),
);
const memoSchema = v.pipe(v.string(), v.trim());

function dateInputSchema(today: string) {
  return v.pipe(
    v.string(),
    v.trim(),
    v.check((date) => {
      if (!datePattern.test(date) || todayJst(new Date(`${date}T00:00:00+09:00`)) !== date) {
        return false;
      }
      return date <= today;
    }, dateError),
  );
}

export function nyushukkinInputSchema(today: string) {
  return v.object({
    kind: kindSchema,
    amount: amountInputSchema,
    date: dateInputSchema(today),
    memo: memoSchema,
  });
}

function fromSchema<TSchema extends v.BaseSchema<unknown, unknown, v.BaseIssue<unknown>>>(
  schema: TSchema,
  input: unknown,
): NyushukkinResult<v.InferOutput<TSchema>> {
  const parsed = v.safeParse(schema, input);
  if (parsed.success) {
    return { ok: true, value: parsed.output };
  }
  return { ok: false, error: parsed.issues[0].message };
}

export function todayJst(now = new Date()): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Tokyo",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(now);
}

export function calendarMonth(date: string): string {
  return date.slice(0, 7);
}

export function shiftCalendarMonth(month: string, delta: number): string {
  const year = Number(month.slice(0, 4));
  const monthIndex = Number(month.slice(5, 7)) - 1 + delta;
  const shifted = new Date(Date.UTC(year, monthIndex, 1));
  const nextYear = shifted.getUTCFullYear();
  const nextMonth = String(shifted.getUTCMonth() + 1).padStart(2, "0");
  return `${nextYear}-${nextMonth}`;
}

export function isCurrentOrPastMonth(month: string, today: string): boolean {
  return month <= calendarMonth(today);
}

export function formatCalendarMonth(month: string): string {
  return `${month.slice(0, 4)}年${Number(month.slice(5, 7))}月`;
}

export function nyushukkinInMonth(items: Nyushukkin[], month: string): Nyushukkin[] {
  return items.filter((item) => calendarMonth(item.date) === month);
}

function parseKind(raw: string): NyushukkinResult<Kind> {
  return fromSchema(kindSchema, raw);
}

function parseDate(raw: string, today: string): NyushukkinResult<string> {
  return fromSchema(dateInputSchema(today), raw);
}

export function recordNyushukkin(
  input: NyushukkinInput,
  today: string,
  id: string,
): NyushukkinResult<Nyushukkin> {
  return assemble(id, input, today);
}

export function correctNyushukkin(
  current: Nyushukkin,
  input: NyushukkinInput,
  today: string,
): NyushukkinResult<Nyushukkin> {
  return assemble(current.id, input, today);
}

export function removeNyushukkin(items: Nyushukkin[], id: string): NyushukkinResult<Nyushukkin[]> {
  if (!items.some((item) => item.id === id)) {
    return { ok: false, error: "その入出金はありません" };
  }
  return { ok: true, value: items.filter((item) => item.id !== id) };
}

export function restoreNyushukkin(
  items: Nyushukkin[],
  removed: Nyushukkin,
): NyushukkinResult<Nyushukkin[]> {
  if (items.some((item) => item.id === removed.id)) {
    return { ok: false, error: "その入出金はすでにあります" };
  }
  return { ok: true, value: [...items, removed] };
}

export function replaceNyushukkin(items: Nyushukkin[], next: Nyushukkin): Nyushukkin[] {
  const index = items.findIndex((item) => item.id === next.id);
  if (index === -1) {
    return [...items, next];
  }
  return items.map((item) => (item.id === next.id ? next : item));
}

export function monthTotals(items: Nyushukkin[], month: string) {
  let income = 0;
  let expense = 0;
  for (const item of items) {
    if (calendarMonth(item.date) !== month) {
      continue;
    }
    if (item.kind === "収入") {
      income += item.amount;
    } else {
      expense += item.amount;
    }
  }
  return { income, expense, balance: income - expense };
}

export function sortNyushukkin(items: Nyushukkin[]): Nyushukkin[] {
  return [...items].sort((a, b) => {
    if (a.date !== b.date) {
      return a.date < b.date ? 1 : -1;
    }
    return a.id < b.id ? 1 : -1;
  });
}

export function readStored(raw: string | null): NyushukkinResult<Nyushukkin[]> {
  if (raw == null) {
    return { ok: false, error: "帳簿がありません" };
  }
  try {
    const parsed: unknown = JSON.parse(raw);
    if (!Array.isArray(parsed)) {
      return { ok: false, error: "帳簿が壊れています" };
    }
    const items: Nyushukkin[] = [];
    for (const row of parsed) {
      const item = asNyushukkin(row);
      if (item) {
        items.push(item);
      }
    }
    return { ok: true, value: items };
  } catch {
    return { ok: false, error: "帳簿が壊れています" };
  }
}

export function loadStored(primary: string | null, backup: string | null): Nyushukkin[] {
  const main = readStored(primary);
  if (main.ok) {
    return main.value;
  }
  const prev = readStored(backup);
  if (prev.ok) {
    return prev.value;
  }
  return [];
}

export function serializeStored(items: Nyushukkin[]): string {
  return JSON.stringify(items);
}

function assemble(id: string, input: NyushukkinInput, today: string): NyushukkinResult<Nyushukkin> {
  const parsed = fromSchema(nyushukkinInputSchema(today), input);
  if (!parsed.ok) {
    return parsed;
  }
  return {
    ok: true,
    value: {
      id,
      kind: parsed.value.kind,
      amount: Number(parsed.value.amount),
      date: parsed.value.date,
      memo: parsed.value.memo,
    },
  };
}

function asNyushukkin(row: unknown): Nyushukkin | undefined {
  if (!row || typeof row !== "object") {
    return undefined;
  }
  const rec = row as Record<string, unknown>;
  if (typeof rec.id !== "string" || rec.id === "") {
    return undefined;
  }
  const kind = parseKind(String(rec.kind ?? ""));
  if (!kind.ok) {
    return undefined;
  }
  if (typeof rec.amount !== "number" || !Number.isSafeInteger(rec.amount) || rec.amount < 1) {
    return undefined;
  }
  const date = parseDate(String(rec.date ?? ""), "9999-12-31");
  if (!date.ok) {
    return undefined;
  }
  if (typeof rec.memo !== "string") {
    return undefined;
  }
  return {
    id: rec.id,
    kind: kind.value,
    amount: rec.amount,
    date: date.value,
    memo: rec.memo,
  };
}

File: api/cmd/server/nyushukkin.go
package main

import (
	"cmp"
	"encoding/json"
	"errors"
	"slices"
	"strconv"
	"strings"
	"time"
)

const (
	kindExpense = "支出"
	kindIncome  = "収入"
	kindError   = "収入か支出を選んでください"
	amountError = "金額は1円以上の整数円です"
	dateError   = "入出日は今日以前の日付です"
	monthError  = "暦月を指定してください"
	monthFormat = "暦月の形式が違います"
	notFound    = "その入出金はありません"
	maxSafe     = 9007199254740991
)

var (
	errNotFound = errors.New(notFound)
	jst         = time.FixedZone("JST", 9*3600)
)

type Nyushukkin struct {
	ID     string `json:"id"`
	Kind   string `json:"kind"`
	Amount int64  `json:"amount"`
	Date   string `json:"date"`
	Memo   string `json:"memo"`
}

type nyushukkinInput struct {
	Kind   string      `json:"kind"`
	Amount json.Number `json:"amount"`
	Date   string      `json:"date"`
	Memo   *string     `json:"memo"`
}

type validateError struct {
	msg string
}

func (e validateError) Error() string {
	return e.msg
}

func todayJST(now time.Time) string {
	return now.In(jst).Format("2006-01-02")
}

func parseKind(raw string) (string, error) {
	if raw == kindExpense || raw == kindIncome {
		return raw, nil
	}
	return "", validateError{kindError}
}

func parseAmount(raw json.Number) (int64, error) {
	s := string(raw)
	if s == "" || strings.ContainsAny(s, ".eE+-") {
		return 0, validateError{amountError}
	}
	v, err := strconv.ParseInt(s, 10, 64)
	if err != nil || v < 1 || v > maxSafe {
		return 0, validateError{amountError}
	}
	return v, nil
}

func parseDate(raw string, today string) (string, error) {
	date := strings.TrimSpace(raw)
	t, err := time.ParseInLocation("2006-01-02", date, jst)
	if err != nil || t.Format("2006-01-02") != date {
		return "", validateError{dateError}
	}
	if date > today {
		return "", validateError{dateError}
	}
	return date, nil
}

func parseMonth(raw string) (string, error) {
	month := strings.TrimSpace(raw)
	if month == "" {
		return "", validateError{monthError}
	}
	t, err := time.ParseInLocation("2006-01", month, jst)
	if err != nil || t.Format("2006-01") != month {
		return "", validateError{monthFormat}
	}
	return month, nil
}

func calendarMonth(date string) string {
	if len(date) < 7 {
		return ""
	}
	return date[:7]
}

func parseInput(input nyushukkinInput, today string) (kind string, amount int64, date string, memo string, err error) {
	kind, err = parseKind(input.Kind)
	if err != nil {
		return "", 0, "", "", err
	}
	amount, err = parseAmount(input.Amount)
	if err != nil {
		return "", 0, "", "", err
	}
	date, err = parseDate(input.Date, today)
	if err != nil {
		return "", 0, "", "", err
	}
	if input.Memo != nil {
		memo = strings.TrimSpace(*input.Memo)
	}
	return kind, amount, date, memo, nil
}

func recordNyushukkin(input nyushukkinInput, today string, id string) (Nyushukkin, error) {
	return assemble(id, input, today)
}

func correctNyushukkin(current Nyushukkin, input nyushukkinInput, today string) (Nyushukkin, error) {
	return assemble(current.ID, input, today)
}

func assemble(id string, input nyushukkinInput, today string) (Nyushukkin, error) {
	kind, amount, date, memo, err := parseInput(input, today)
	if err != nil {
		return Nyushukkin{}, err
	}
	return Nyushukkin{ID: id, Kind: kind, Amount: amount, Date: date, Memo: memo}, nil
}

func removeNyushukkin(items []Nyushukkin, id string) ([]Nyushukkin, Nyushukkin, error) {
	next := make([]Nyushukkin, 0, len(items))
	var removed Nyushukkin
	found := false
	for _, item := range items {
		if item.ID == id {
			removed = item
			found = true
			continue
		}
		next = append(next, item)
	}
	if !found {
		return nil, Nyushukkin{}, errNotFound
	}
	return next, removed, nil
}

func nyushukkinInMonth(items []Nyushukkin, month string) []Nyushukkin {
	out := make([]Nyushukkin, 0)
	for _, item := range items {
		if calendarMonth(item.Date) == month {
			out = append(out, item)
		}
	}
	return out
}

func sortNyushukkin(items []Nyushukkin) []Nyushukkin {
	out := slices.Clone(items)
	slices.SortFunc(out, func(a, b Nyushukkin) int {
		if a.Date != b.Date {
			return cmp.Compare(b.Date, a.Date)
		}
		return cmp.Compare(b.ID, a.ID)
	})
	return out
}

func asNyushukkin(row any) (Nyushukkin, bool) {
	rec, ok := row.(map[string]any)
	if !ok {
		return Nyushukkin{}, false
	}
	id, _ := rec["id"].(string)
	if id == "" {
		return Nyushukkin{}, false
	}
	kind, err := parseKind(asString(rec["kind"]))
	if err != nil {
		return Nyushukkin{}, false
	}
	amount, ok := asStoredAmount(rec["amount"])
	if !ok {
		return Nyushukkin{}, false
	}
	date, err := parseDate(asString(rec["date"]), "9999-12-31")
	if err != nil {
		return Nyushukkin{}, false
	}
	memo, ok := rec["memo"].(string)
	if !ok {
		return Nyushukkin{}, false
	}
	return Nyushukkin{ID: id, Kind: kind, Amount: amount, Date: date, Memo: memo}, true
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}

func asStoredAmount(v any) (int64, bool) {
	n, ok := v.(float64)
	if !ok || n != float64(int64(n)) || n < 1 || n > maxSafe {
		return 0, false
	}
	return int64(n), true
}
